include/revng/Model/Type.h

```diff
@@ -0,0 +1,699 @@
+  std::string Name;
+
+  AggregateField(const QualifiedType &QT, llvm::StringRef NameRef) :
+    FieldType(QT), Name(NameRef) {}
+
+  AggregateField(QualifiedType &&QT, llvm::StringRef NameRef) :
+    FieldType(std::move(QT)), Name(NameRef) {}
+
+  AggregateField(const QualifiedType &QT) : AggregateField(QT, "") {}
+
+  AggregateField(QualifiedType &&QT) : AggregateField(std::move(QT), "") {}
+
+  AggregateField() :
+    AggregateField(QualifiedType{ TypePath::fromString("/"), {} }) {}
+
+  bool operator==(const AggregateField &) const = default;
+};
+
+/// \brief A field of a struct type in model, with offset, qualified type, and
+/// name
+struct model::StructField : model::AggregateField {
+  uint64_t Offset;
+
+  StructField(uint64_t O, llvm::StringRef NameRef, const QualifiedType &QT) :
+    AggregateField(QT, NameRef), Offset(O) {}
+
+  StructField(uint64_t O) : StructField(O, "", {}) {}
+
+  StructField() : StructField(0) {}
+
+  bool operator==(const StructField &) const = default;
+};
+INTROSPECTION_NS(model, StructField, Offset, Name, FieldType);
+
+/// \brief KeyedObjectTraits for model::StructType based on its byte-offset in
+/// struct
+template<>
+struct KeyedObjectTraits<model::StructField> {
+
+  static uint64_t key(const model::StructField &Val) { return Val.Offset; }
+
+  static model::StructField fromKey(const uint64_t &Offset) {
+    return model::StructField(Offset);
+  }
+};
+
+/// \brief Make StructField yaml-serializable
+template<>
+struct llvm::yaml::MappingTraits<model::StructField>
+  : public TupleLikeMappingTraits<model::StructField> {};
+
+/// \brief A struct type in model. Structs are actually typedefs of unnamed
+/// structs in C.
+struct model::StructType : public model::Type {
+  static constexpr const char *Tag = "!Struct";
+  static constexpr const TypeKind::Values AssociatedKind = TypeKind::Struct;
+
+  SortedVector<StructField> Fields;
+  uint64_t Size;
+
+  StructType(llvm::StringRef NameRef) :
+    Type(TypeKind::Struct, NameRef), Fields() {}
+
+  StructType() : StructType("") {}
+
+  static bool classof(const Type *T) { return T->Kind == TypeKind::Struct; }
+
+  // Not to be used directly, only KeyedObjectTraits should use this
+  StructType(uint64_t ID) : Type(TypeKind::Struct, ID, "") {}
+};
+INTROSPECTION_NS(model, StructType, Kind, ID, Name, Fields, Size);
+
+/// \brief A field of a union type in model, with position, qualified type, and
+/// name
+struct model::UnionField : model::AggregateField {
+  uint64_t ID;
+
+public:
+  UnionField(const QualifiedType &QT, llvm::StringRef NameRef);
+
+  UnionField(QualifiedType &&QT, llvm::StringRef NameRef);
+
+  UnionField() : UnionField({}, "") {}
+
+private:
+  UnionField(uint64_t TheID, QualifiedType &&QT, llvm::StringRef NameRef) :
+    AggregateField(std::move(QT), NameRef), ID(TheID) {}
+
+public:
+  UnionField(uint64_t TheID) : UnionField(TheID, {}, "") {}
+
+  bool operator==(const UnionField &Other) const { return ID == Other.ID; }
```

Comment:
Why?

Reply:
Leftover, it should be `= default` like for `StructType`.